src/can.rs

```diff
@@ -0,0 +1,181 @@
+//! Controller Area Network
+
+/// A CAN2.0 Frame
+pub trait Frame {
+    /// Creates a new frame with a standard identifier.
+    fn new_standard(id: u32, data: &[u8]) -> Self;
+
+    /// Creates a new frame with an extended identifier.
+    fn new_extended(id: u32, data: &[u8]) -> Self;
+
+    /// Marks this frame as a remote frame (by setting the RTR bit).
+    fn with_rtr(&mut self, dlc: usize) -> &mut Self;
+
+    /// Returns true if this frame is a extended frame.
+    fn is_extended(&self) -> bool;
+
+    /// Returns true if this frame is a standard frame.
+    fn is_standard(&self) -> bool {
+        !self.is_extended()
+    }
+
+    /// Returns true if this frame is a remote frame.
+    fn is_remote_frame(&self) -> bool;
+
+    /// Returns true if this frame is a data frame.
+    fn is_data_frame(&self) -> bool {
+        !self.is_remote_frame()
+    }
+
+    /// Returns the frame identifier.
+    fn id(&self) -> u32;
+
+    /// Returns the data length code (DLC) which is in the range 0..8.
+    ///
+    /// For data frames the DLC value always matches the lenght of the data.
+    /// Remote frames no not carry any data, yet the DLC can be greater than 0.
+    fn dlc(&self) -> usize;
+
+    /// Returns the frame data (0..8 bytes in length).
+    fn data(&self) -> &[u8];
+}
+
+/// A CAN interface that is able to transmit frames.
+pub trait Transmitter {
+    /// Associated frame type.
+    type Frame: Frame;
```

Comment:
Is this going to work for device crates? (be advised that I may very well be off)

Here's what happens when I tried doing so:

1. Because it's a device crate, there's no transmitter implementation, I just use the traits like this:
```
pub struct Device<T: Receiver + Transmitter> {
    bus: T
}
```
2. When I go to use the 'Transmitter::transmit' function, I only have a `Frame` trait to work with so I wrote a `CanFrame` struct that implements the `Frame` trait.
3. Problem is I can't for the life of me pass my `CanFrame` struct to the `transmit` function.  Here's what my implementation looks like: (cropped for readability) 
```
impl<T, E, F> Device<T>
where
    T: Receiver<Error = E, Frame = F> + Transmitter<Error = E, Frame = F>,
    F: crate::hal::can::Frame,
    E: Debug,
{
...
```

Note also that I tried creating a Mock Transmitter and I could include the types and that works fine:
```
    impl Transmitter for MockCanBus {
        type Frame = CanFrame;
        type Error = ();

        fn transmit(&mut self, frame: &CanFrame) -> nb::Result<Option<Self::Frame>, Self::Error> {
            self.frames.push(frame.clone());
            Ok(Option::None)
        }
    }
```

Is there problem between the chair and the keyboard?

Reply:
> Is this going to work for device crates?

This is a very valid question! Actually I started this PR because the previous attempts for CAN traits failed to realize this.

For embedded-hal traits there are usually two perspectives:
1. HAL implementations (like `stm32f1xx-hal` or `lpc8xx-hal`) which actually provide and have the `impl Trait for xxx`.
2. Device drivers which only consume the trait interfaces (but do not implement them)

Now to your points regarding the implementation of a device driver:
```` rust
pub struct Device<T: Receiver + Transmitter> {
    bus: T
}
````
This will work fine for crates that implement the `Receiver` and `Transmitter` trait on the same type. This might not always be the case. To make the driver work in the general I recommend following code (note how the traits bound can also be omitted here):
``` rust
struct Device<Rx, Tx> {
    rx: Rx,
    tx: Tx,
}
````

> When I go to use the 'Transmitter::transmit' function, I only have a Frame trait to work with so I wrote a CanFrame struct that implements the Frame trait

Now comes the tricky part and you were almost there. Crucial point is that you do not need to include the `Frame` in the generic parameter list. To use the `Frame` trait you access it through the `Tx` type parameter instead like `Tx::Frame`.

```` rust
impl<Tx, ...> Device<Tx, ...>
where
    Tx: Transmitter,
    ...
{
    pub fn example() {
        let tx_frame = Tx::Frame::new_standard(0x123, &[]);
        block!(self.tx.transmit(&tx_frame)).unwrap();
    }
}
````

Hope that clarifies some things!

For details you can have a look at this [example device driver](https://github.com/timokroeger/pcan-basic-rs/blob/6a4868bb341ff1c5807a77e484e01cda02398352/pcan-basic/examples/stm32-fwupdate.rs).

Reply:
Yep, that works.  Thanks a million Timo for the heads up.  Very elegant approach I must say (once you wrapped your head around it).

It might be a good idea to note this in the `Frame`'s and the associated type in `Transmitter` (above `type Frame: Frame;`) documentation?  One is usually tempted to go straight to what he needs, but in this case, there's a little detour to take... something tells me I won't be the only one to make this mistake.

FYI: I'm going to take a stab at implementing a Linux implementation.  Only thing I found out there is socketcan-rs, but that's terribly outdated and includes too many dependencies for my liking.

Again, compliments on your implementation, very nice work

Reply:
Made some progress.  I have a first draft of a Linux implementation loosely based on socketcan-rs that only uses libc here: https://github.com/reneherrero/candev

A couple of comments:
1. Would it not be a good idea to have `Frame::new_standard` and `Frame::new_extended` return a result?  Unless I'm missing something, as it stands, it is possible to pass and incorrect id or too much data?
2. I made a few examples to test things out and can transmit and receive data fine using Timo's traits.  But when I try to implement a trivial [driver](https://github.com/reneherrero/candev/blob/master/examples/driver.rs), it does not recognize that the `Receiver::Frame` and the `Transmitter::Frame` are the same type and hence doing the same as @timokroeger does in the `stm32f1xx-hal` `echo` example gives me an `mismatched types` error.  Not clear on the why, but pretty sure it's something I'm doing wrong.
```
if let Ok(frame) = self.t.receive() {
  self.t.transmit(&frame);
                  ^^^^^^ expected embedded_hal::can::Transmitter::Frame, found embedded_hal::can::Receiver::Frame
}
```

Thoughts?

Reply:
Good to see more implementations to verify and improve the traits!

1. Pushed a commit to fix this. As noted previously I do not feel like it makes sense to recover from an error during frame creation. That’s why I did not add an additional `Error` associate type. I opted for a simple `()` error case to keep complexity low.

2. To make this work rust needs to support https://github.com/rust-lang/rust/issues/20041. I have think more about this one. Maybe merging receiver and transmitter traits makes sense after all.

Reply:
Took a stab at the filters yesterday.

Implementing the `Frame` trait went fairly well: just need to do some testing to ensure correctness...  the Soketcan documentation is kind of bare.

Where I did run into a wall is with the `FilteredReceiver`: Socketcan doesn't seem to have the notion of filter groups and I'm not sure how to implement the 'FilterGroup' and `FilterGroups` traits along with the `filter_groups` method which 'FilteredReceiver' needs.

Multiple ways to see this:
1. Would be possible to just not implement `FilteredReceiver` then create and configure the filters on the Socket object and pass it along to whatever.  Problem with that is that anything using `FilteredReceiver` wouldn't work: not good.
2. Other is to just stub it out the parts that don't make sense which would at least enable creating filters directly with the hal.  Better than (1) but still not great.
3. Another I can think of is a blend of (1) and (2), separating the filter group pieces from `FilteredReceiver`.

Tough call.  I hate the fact that looking for the lowest common denominator often results in less than optimal results.

Any suggestions?  Are partial implementations an ok thing?

Reply:
For time reasons only abrief comments. I’ll revisit later.

SocketCAN handles filtering in SW so theoretically it has no limit on the number of filters.  
When all filters have the same capability (which I think they have for SocketCAN) then there is only one `FilterGroup`.  
Because there is just one `FilterGroup` you can simple use `iter::Once<FilterGroup>` for the `FilterGroups` associate type  and return `iter::once(…)` from the `filter_groups()` method.

Reply:
I understand.  I looked at your stm32f1xx-hal implementation for inspiration but couldn't see how to translate it to SocketCAN.

In the meantime I added a basic CI, docs and updated the dependencies to use your latest changes.  Will ponder on the idea...

Thanks Timo

Reply:
Ref: https://www.kernel.org/doc/html/v5.4/networking/can.html#raw-protocol-sockets-with-can-filters-sock-raw

The filter group trait implementation for SocketCAN can look like:
```` rust
pub struct FilterGroup;

impl embedded_hal::can::FilterGroup for FilterGroup {
    fn num_filters(&self) -> usize {
        // CAN_RAW_FILTER_MAX
        // https://github.com/torvalds/linux/blob/v5.4/include/uapi/linux/can.h#L218
        512
    }

    fn extended(&self) -> bool {
        // Filter works for extended (29bit) identifiers
        true
    }

    fn mask(&self) -> Option<MaskType> {
        // There is a configurable mask for each filter
        Some(MaskType::Individual)
    }

    fn rtr(&self) -> RtrFilterBehavior {
        // RTR bit is part of the the filter and the mask
        RtrFilterBehavior::Configurable
    }
}

impl embedded_hal::can::FilteredReceiver for SocketCanDevice {
    type Filter = Filter;
    type FilterGroup = FilterGroup;
    type FilterGroups = std::iter::Once<FilterGroup>;

    fn filter_groups(&self) -> Once<FilterGroup> {
        // All filters have the same capability: There is only one group.
        std::iter::once(FilterGroup)
    }

    // ...
}
````

Reply:
Found another error in the filter description and added a commit to clarify.
Thank you for implementing the traits for SocketCAN! It really helps to polish the interfaces.

Reply:
Might also help to look at the [Broadcast Manager](https://www.kernel.org/doc/Documentation/networking/can.txt) part for the filtering part. I would still split filters and basic receive send the faster we get the basic stuff the better, I think that already makes a lot of things possible.

Reply:
No problem Timo... thank you for your patience/tolerance, I'm getting pretty decent at rust, but still learning.  Same goes for the CAN echo system.

I put all your suggestions in place, although I need to read up some more on filter groups to make sure what I put in did mimics well what's at the hardware level.  Mind you that changes would be internal and not break anything.

Ok, still needs quite a bit of testing and polishing (don't like the candev name either -> suggestions welcome), but it is a decent start I think.

Reply:
There is one thing I should have mentioned, and it only happens when implementing a driver: using the `unwrap` on `transmit` throws a compile error.  Tried to fix this multiple ways but no luck.  Works just fine on the receive and transmit examples.
```
no method named `unwrap` found for enum `std::result::Result<std::option::Option<<T as embedded_hal::can::Transmitter>::Frame>, nb::Error<<T as embedded_hal::can::Transmitter>::Error>>` in the current scope

method not found in `std::result::Result<std::option::Option<<T as embedded_hal::can::Transmitter>::Frame>, nb::Error<<T as embedded_hal::can::Transmitter>::Error>>`

note: the method `unwrap` exists but the following trait bounds were not satisfied:
      `nb::Error<<T as embedded_hal::can::Transmitter>::Error>: std::fmt::Debug`
```

Thoughts?  Line of code is [here](https://github.com/reneherrero/candev/blob/167a201f42963448f7f93c38ee617ca5a22c1519/examples/driver.rs#L26)

Reply:
That is due to the nature of [`nb::Result`](https://docs.rs/nb/1.0.0/nb/index.html).  
`block!(tx.transmit(&frame)).unwrap()` should work.

Reply:
Tried that too: doesn't work either.
```
block!(self.t.transmit(&frame)).unwrap();
   |                            ^^^^^^ method not found in `std::result::Result<std::option::Option<<T as embedded_hal::can::Transmitter>::Frame>, <T as embedded_hal::can::Transmitter>::Error>`
   |
   = note: the method `unwrap` exists but the following trait bounds were not satisfied:
           `<T as embedded_hal::can::Transmitter>::Error: std::fmt::Debug`
```

I did finally find a way to fix this, but not sure it's the right way of going about it: by changing `type Error` in [`Transmiter`](https://github.com/timokroeger/embedded-hal/blob/2c935c1e2c1c6229ef3fc590da519a13b90faae6/src/can.rs#L51) to `type Error: core::fmt::Debug;`

You can reproduce by creating an `examples` folder in your copy of `embedded-hal` and adding an `rs` file with this:
```
use embedded_hal::can::{Frame, Receiver, Transmitter};
use nb::block;

#[derive(Debug)]
struct Driver<T> {
    pub id: u32,
    pub t: T,
}

impl<T> Driver<T>
where
    T: Receiver + Transmitter,
{
    pub fn echo(&mut self) {
        let mut count = 0;
        while count < 3 {
            /* This doesn't work:
            if let Ok(frame) = self.t.receive() {
                self.t.transmit(&frame);
            }
            */
            if let Ok(frame) = self.t.receive() {
                let frame = <T as Transmitter>::Frame::new_standard(self.id, frame.data()).unwrap();
                block!(self.t.transmit(&frame)).unwrap();
                // match self.t.transmit(&frame) {
                //     Ok(f) => match f {
                //         None => {}
                //         Some(_) => panic!("Transmit returned a queued frame"),
                //     },
                //     Err(_) => panic!("Failed to transmit frame"),
                // }
            }

            count += 1;
        }
    }
}

fn main(){}
```

Reply:
For `unwrap()` to work you need to add the `Debug` trait bound for `Transmitter::Error` ([example](https://github.com/timokroeger/pcan-basic-rs/blob/master/pcan-basic/examples/stm32-fwupdate.rs#L15-L22)).
To keep things simple your commented out manual error handling is perfectly fine.

Reply:
Thanks to pointing me to that example, got a variation of that to work:
```
impl<T> Driver<T>
where
    T: Receiver + Transmitter,
    <T as Transmitter>::Error: core::fmt::Debug,
{ ... }
```

But to your point, any real world scenario probably wouldn't use `unwrap`, or send back the same exact message it received for that matter.

Hope this helps move this PR forward: looking pretty solid from where I'm standing.

@ilya-epifanov Can you comment?

Reply:
@reneherrero
Found more elegant way to achieve frame loopback which actually works on stable:
```` rust
impl<T, F, E> Driver<T>
where
    F: Frame,
    E: core::fmt::Debug,
    T: Receiver<Frame = F, Error = E> + Transmitter<Frame = F, Error = E>,
{
    pub fn echo(&mut self) {
        let frame = block!(self.0.receive()).unwrap();
        self.0.transmit(&frame).unwrap();
    }
}
````

Reply:
WOW!  Thanks for the Associated Types 301 class @timokroeger 🥇